### ISSUE TITLE
Backport Moondust's background rendering engine

### ISSUE THREAD

**Issue by Wohlstand** [Wohlstand/TheXTech#463]
This might be a sub-task of task #272, however, it can work by itself, just with a set of limitations: you won't be able to render elements between in-scene objects, but should be able to render elements at back, and at front of the thing. Background-2.ini files will be fully based on Moondust's implementation and they won't match SMBX2's implementation.

**Comment by LoveBodhi** [comment]
This is just related in my issue #94 .

**Comment by LoveBodhi** [comment]
I should suggest "color" field, for blend colour.

**Comment by ds-sloth** [comment]
@Wohlstand, moving to Version 1.3.7.

**Comment by Wohlstand** [comment]
Ye, moving to 1.3.7, as I said before, it's better to minify new features as much as possible as we already had delayed the 1.3.6.1 for a while.

**Comment by LoveBodhi** [comment]
I wonder does this can be attached with **Shader** system...

**Comment by Wohlstand** [comment]
No, the X2 tag is not for here. The multi-layer backgrounds system is different than X2, and it has different specification of INI files which is somewhere incompatible.

**Comment by ds-sloth** [comment]
I do think we should consider simplifying the background specification from the draft on the Moondust wiki, but we shouldn't aim for total compatibility with the X2 specification. The X2 specification still seems too complex, but it might be the case that anything simpler could not reproduce all SMBX 1.3 backgrounds.

**Comment by Wohlstand** [comment]
Actually, Moondust's specification has two variants of backgrounds:
- Single-layer which exists since foundation of Moondust, it has designed to fully reproduce all SMBX 1.3's backgrounds and have the united mechanism that does that
- Multi-layer which allows making more advanced backgrounds, and it has been implemented as a separated class, so, both variants supported and works.

Also the fact: Hoeloe was totally against the single-layer backgrounds module and even requested me to have it being "legacy" and hide the ability to use it. But that was not my with at all, I wanted both systems to work especially because the single-layer background also has the "stripes" mechanism that allows to chop the whole canvas into stripes that has different paralax coefficients. Hoeloe, as usually, disliked this concept at all. Anyway, if do think about layers, it's WAY more complicated to build the similar stripped canvas by layers, making separated layer for each stripe and specifying different relative offsets manually. In a case of my implementation, I have to specify an array of cut-lines offsets and the array of paralax coefficients for every strip.

**Comment by ds-sloth** [comment]
I actually agree here, in the data structure it's better for every layer to have only a single strip. This way, there is just one array of layers, and no need for "sub strips" inside a single layer.

I don't know whether it should be done this way in the INI as well, but it will be very helpful for the memory usage to do this in the struct itself.

**Comment by Wohlstand** [comment]
In a case of my implementation, the "stripped" background has one signle texture and arrays of cut offsets and paralax coefficients, and every strip gets drawn separately using given texture offsets. That how it works. If these stripes would be a separated layers, they will be a separated physical pictures (in the case if we don't have a choice to crop the texture and re-use the same texture in different layers).

**Comment by LoveBodhi** [comment]
Maybe after Online Functionality?

**Comment by ds-sloth** [comment]
That's up to Wohlstand who will be implementing this feature in parallel with my work on online functionality.

I will probably not personally help with it until after I've implemented NetPlay.

**Comment by Wohlstand** [comment]
The netplay work right now is done by ds-sloth, this task is mine, and I going to perform it after completing the pile of other tasks

**Comment by ChristianSilvermoon** [comment]
Would this enable the easier creation of things like foreground fog?
